Handle photometry in both flux and magnitude space

dannygoldstein opened the pull request:
This PR introduces some significant changes to the `skyportal` Photometry API and database data model:

   * The Photometry API now allows you to POST photometry in both flux and magnitude space: https://portal.nersc.gov/cfs/m937/openapi2.html#/paths/~1api~1photometry/post, including a mixture of both types in the same request

<img width="979" alt="Screen Shot 2020-05-28 at 12 42 31 AM" src="https://user-images.githubusercontent.com/2769632/83099763-31327280-a07c-11ea-900d-540ddcfd542e.png">
<img width="982" alt="Screen Shot 2020-05-28 at 12 42 36 AM" src="https://user-images.githubusercontent.com/2769632/83099764-31327280-a07c-11ea-99f9-0ec577662aa7.png">
<img width="984" alt="Screen Shot 2020-05-28 at 12 42 40 AM" src="https://user-images.githubusercontent.com/2769632/83099765-31cb0900-a07c-11ea-9a52-2c4f4b4e652b.png">

   * The Photometry API now allows you to GET photometry in either flux or magnitude space (check the 200 response): https://portal.nersc.gov/cfs/m937/openapi2.html#/paths/~1api~1photometry~1photometry_id/get

   * The Photometry SQLa model now stores the original packet that was used to pass the photometry 

   * The Thumbnail API has been separated from the Photometry API

   * The meat of the handlers has been simplified and error handling has been significantly improved using Marshmallow deserialization 

   * Filter checking is performed on incoming photometry points to make sure the filters are compatible with the proposed instruments 

   * All fluxes are now stored in the database in microJanskies in the AB system

dannygoldstein commented:
[ch916]

dannygoldstein commented:
Ready for review

acrellin commented on skyportal/tests/api/test_photometry.py:
```diff
@@ -23,49 +27,193 @@ def test_token_user_post_get_photometry_data(upload_data_token, public_source):
+    assert status == 400
+    assert data['status'] == 'error'
+
+    # invalid request
```
Valid?

acrellin commented:
Are we sure we want to require passing a list of objects that each duplicate a lot of the same data (object ID, filter, etc) in order to post a list of photometry points? This feels much more cumbersome than passing a single dict with mag/flux keys whose values are lists. We could require each request only contain one type of data. Thoughts?

stefanv commented:
> Are we sure we want to require passing a list of objects that each duplicate a lot of the same data (object ID, filter, etc) in order to post a list of photometry points? This feels much more cumbersome than passing a single dict with mag/flux keys whose values are lists. We could require that each request only contain one type of data. Thoughts?

I'd lean in that direction, too.

dannygoldstein commented:
Hi both, sure, I can put it back.

dannygoldstein commented:
There is also something nice about having each photometry point logically grouped into a single unit, instead of the fields of a single point spread across multiple lists, so you always know what goes together. Makes debugging easier. With these light weight packets, the extra overhead of the duplicated keys is pretty much negligible from a computational / data processing perspective, so the issue is primarily one of syntactic sugar - presumably people could alleviate this with list comprehension in generating the packets so that they only have to type each key one time

dannygoldstein commented:
Before I revert back to the old way, curious to hear your thoughts @stefanv and @acrellin  on my comment above

acrellin commented:
I'm still leaning towards ease of use for the API user -- having to generate the list of packets that each contain all of the relevant fields feels like a barrier that'd make me less inclined to use the API (and also seems more error-prone), whereas just passing lists of mags, errs, times, to the relevant fields and just having to specify everything else once the request data feels simpler and easier from a user-perspective.

acrellin commented:
We could also require homogeneous data types for each request so that some other fields could perhaps become a single number rather than having to specify a list (if they're typically all the same in a given request)

acrellin commented:
@profjsb @stefanv what are your thoughts on the above?

dannygoldstein commented:
> We could also require homogeneous data types for each request so that some other fields could perhaps become a single number rather than having to specify a list (if they're typically all the same in a given request)

Is there an easy way to do this in `marshmallow`?  I think this could quickly become complex to capture in the docs if we have, e.g., `mag = fields.Number()` and a separate schema with `mag = fields.List(fields.Number())` and similar for the other fields and combinations of fields (the combinatorial possibilities quickly become large)

acrellin commented:
Those we could require to always be lists

acrellin commented:
Well, we already handle the data sometimes being a list and sometimes not, in which case we just wrap the single datum in a list

dannygoldstein commented:
> Those we could require to always be lists

I think this can be a bit more error prone from a validation perspective as it introduces some funky broadcasting behavior - take for example the situation where someone passes `data = {'mag': [19.], 'magerr': [0.1, 0.2]}`. Perhaps they meant `'mag': [19., 20.]` but forgot to include the `20.`, then the mag of their second element is silently set to 20 via broadcasting, whereas if mag could be a scalar, this situation could raise an error and these situations could be avoided

dannygoldstein commented:
Photometry is so important that, I think erring on the side of caution with a slightly more verbose (but concrete) API that keeps data points grouped within logical units is worth it to avoid costly errors down the line

stefanv commented:
@dannygoldstein It would be quite laborious to convert NumPy arrays to the format you propose. The broadcasting issue can be managed by asking NumPy to do it---it knows how to check whether multiple arrays have compatible shapes.  Docs: it is easy to override a single field to mention it can also be a scalar.

If we build two endpoints, `/api/photometry` and `/api/photometry/batch` or some such, we can factor out much of the common code and provide these separately?

dannygoldstein commented:
@stefanv With respect to broadcasting, my view is that the automatic broadcasting is something we would like to avoid, not facilitate, due to the high potential for unintended broadcasting consequences.

These two approaches require the same amount from the user. Example: assume I have a photometry data file `phot.csv` with columns `flux,fluxerr,mjd,filter,zp,magsys`. To facilitate the object of lists approach I would do:

Object of lists (current):

```python
data = np.genfromtxt('phot.csv', dtype=None, names=True, delimiter=',')
object = {key: data[key] for key in data.dtype.names}
# {'filter': array([b'ztfg', b'ztfr', b'ztfg', b'ztfr', b'ztfr'], dtype='|S4'),
 'flags': array([2050, 2050, 2050, 2050, 2050]),
 'flux': array([294.60620237, 274.69987726, 402.67093471, 298.45363937,
        360.49031262]),
 'fluxerr': array([28.94672495, 52.56111254, 53.42758621, 50.645624  , 35.54387586]),
 'id': array([187120429, 186931281, 187260127, 187185822, 187066356]),
 'lim_mag': array([20.4846992, 20.1616001, 19.8264008, 20.1490002, 20.6075993]),
 'mjd': array([58670.2035185, 58677.2021875, 58676.1846528, 58676.1974421,
        58667.2254282]),
 'zp': array([25.87031   , 25.91812899, 25.93607101, 25.916287  , 25.958698  ]),
 'zpsys': array([b'ab', b'ab', b'ab', b'ab', b'ab'], dtype='|S2')}
```

List of objects (proposed):
```python
data = np.genfromtxt('phot.csv', dtype=None, names=True, delimiter=',')
object = [dict(zip(data.dtype.names, row)) for row in data]
#  [{'filter': b'ztfg',
  'flags': 2050,
  'flux': 294.60620237454,
  'fluxerr': 28.9467249460001,
  'id': 187120429,
  'lim_mag': 20.4846992,
  'mjd': 58670.2035185001,
  'zp': 25.870310000000003,
  'zpsys': b'ab'},
 {'filter': b'ztfr',
  'flags': 2050,
  'flux': 274.699877260692,
  'fluxerr': 52.5611125432933,
  'id': 186931281,
  'lim_mag': 20.1616001,
  'mjd': 58677.2021875,
  'zp': 25.91812899,
  'zpsys': b'ab'},
 {'filter': b'ztfg',
  'flags': 2050,
  'flux': 402.670934705762,
  'fluxerr': 53.4275862112982,
  'id': 187260127,
  'lim_mag': 19.8264008,
  'mjd': 58676.1846528002,
  'zp': 25.93607101,
  'zpsys': b'ab'},
 {'filter': b'ztfr',
  'flags': 2050,
  'flux': 298.453639365695,
  'fluxerr': 50.6456239971251,
  'id': 187185822,
  'lim_mag': 20.1490002,
  'mjd': 58676.1974420999,
  'zp': 25.916287,
  'zpsys': b'ab'},
 {'filter': b'ztfr',
  'flags': 2050,
  'flux': 360.490312621869,
  'fluxerr': 35.5438758601201,
  'id': 187066356,
  'lim_mag': 20.6075993,
  'mjd': 58667.2254282003,
  'zp': 25.958698,
  'zpsys': b'ab'}]
```

stefanv commented:
> @stefanv With respect to broadcasting, my view is that the automatic broadcasting is something we would like to avoid, not facilitate, due to the potential for unintended broadcasting consequences.
> 
> These two approaches require the same amount from the user. Example: assume I have a photometry data file `phot.csv` with columns `flux,fluxerr,mjd,filter,zp,zpsys,flags,lim_mag,id`. To facilitate the object of lists approach I would do:
> 
> Object of lists (current):
> 
> ```python
> data = np.genfromtxt('phot.csv', dtype=None, names=True, delimiter=',')
> object = {key: data[key] for key in data.dtype.names}
> # {'filter': array([b'ztfg', b'ztfr', b'ztfg', b'ztfr', b'ztfr'], dtype='|S4'),
>  'flags': array([2050, 2050, 2050, 2050, 2050]),
>  'flux': array([294.60620237, 274.69987726, 402.67093471, 298.45363937,
>         360.49031262]),
>  'fluxerr': array([28.94672495, 52.56111254, 53.42758621, 50.645624  , 35.54387586]),
>  'id': array([187120429, 186931281, 187260127, 187185822, 187066356]),
>  'lim_mag': array([20.4846992, 20.1616001, 19.8264008, 20.1490002, 20.6075993]),
>  'mjd': array([58670.2035185, 58677.2021875, 58676.1846528, 58676.1974421,
>         58667.2254282]),
>  'zp': array([25.87031   , 25.91812899, 25.93607101, 25.916287  , 25.958698  ]),
>  'zpsys': array([b'ab', b'ab', b'ab', b'ab', b'ab'], dtype='|S2')}
> ```
> 
> List of objects (proposed):
> 
> ```python
> data = np.genfromtxt('phot.csv', dtype=None, names=True, delimiter=',')
> object = [dict(zip(data.dtype.names, row)) for row in data]
> #  [{'filter': b'ztfg',
>   'flags': 2050,
>   'flux': 294.60620237454,
>   'fluxerr': 28.9467249460001,
>   'id': 187120429,
>   'lim_mag': 20.4846992,
>   'mjd': 58670.2035185001,
>   'zp': 25.870310000000003,
>   'zpsys': b'ab'},
>  {'filter': b'ztfr',
>   'flags': 2050,
>   'flux': 274.699877260692,
>   'fluxerr': 52.5611125432933,
>   'id': 186931281,
>   'lim_mag': 20.1616001,
>   'mjd': 58677.2021875,
>   'zp': 25.91812899,
>   'zpsys': b'ab'},
>  {'filter': b'ztfg',
>   'flags': 2050,
>   'flux': 402.670934705762,
>   'fluxerr': 53.4275862112982,
>   'id': 187260127,
>   'lim_mag': 19.8264008,
>   'mjd': 58676.1846528002,
>   'zp': 25.93607101,
>   'zpsys': b'ab'},
>  {'filter': b'ztfr',
>   'flags': 2050,
>   'flux': 298.453639365695,
>   'fluxerr': 50.6456239971251,
>   'id': 187185822,
>   'lim_mag': 20.1490002,
>   'mjd': 58676.1974420999,
>   'zp': 25.916287,
>   'zpsys': b'ab'},
>  {'filter': b'ztfr',
>   'flags': 2050,
>   'flux': 360.490312621869,
>   'fluxerr': 35.5438758601201,
>   'id': 187066356,
>   'lim_mag': 20.6075993,
>   'mjd': 58667.2254282003,
>   'zp': 25.958698,
>   'zpsys': b'ab'}]
> ```
> 
> Both approaches require the same amount of code from the user. My view is that the second approach is more concrete, and by logically grouping the data by data point (instead of by field), we can better ensure the integrity of the photometry data we collect from the API

If you start with a CSV file, your data is already in the format you propose.  Consider the situation where a user reads in only certain values from a file, and knows others:

```
filter = 'ztfg'
zpsys = 'ab'
flags = 2050
flux, fluxerr = ... some file io operation ...
```

To fit this into your API, you would have to repeat filter, etc. to match the length of the flux and fluxerr.

Nothing is too hard to do by hand, but we're just thinking what kinds of scenarios may be most commonly encountered.

If you tell me we're basically providing an interface to upload CSV data where all of those columns are always present, then I don't object to the proposed API.

dannygoldstein commented:
@kbarbary any thoughts on this issue?

acrellin commented:
Agreed, if all of the relevant columns (including filter, obj id, zp, zpsys, etc) would typically be present in a csv file and we just want to facilitate posting that, then this is fine. If, on the other hand, the csv file typically contains just a few ofthe  required columns (mjd, mag, err, etc), then the original format is preferred over the one proposed here.

acrellin commented:
In looking at some API design articles, I can't find a single instance of the top-level request JSON structure being a list of objects, but can find plenty where it's an object that contains lists.

acrellin commented:
In reading more about JSON format best practices, the top-level format should be an object rather than an array of objects

acrellin commented:
See, e.g., Google's JSON style guide https://google.github.io/styleguide/jsoncstyleguide.xml#JSON_Structure_&_Reserved_Property_Names

dannygoldstein commented:
OK, object of lists it is. I'll invert the ordering and push up a fix

On Thu, May 28, 2020 at 4:24 PM Ari Crellin-Quick <notifications@github.com>
wrote:

> See, e.g., Google's JSON style guide
> https://google.github.io/styleguide/jsoncstyleguide.xml#JSON_Structure_&_Reserved_Property_Names
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/skyportal/skyportal/pull/472#issuecomment-635587287>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAVEFYF36GAA55FHOC7BNYDRT3B6PANCNFSM4NMXMPJA>
> .
>


-- 
Danny Goldstein
http://astro.caltech.edu/~danny

acrellin commented:
Thanks, @dannygoldstein!

dannygoldstein commented:
@acrellin @stefanv how do you guys recommend handling missing data in this scenario? Specifically, in Flux space, for a given observation, flux can be null, but all the other fields need values. in mag space, both mag and magerr must be null or not null for a given row, and the other values are all required. In the array of objects implementation, this is easy to capture in the docs: for each array element, flux is nullable, and mag and magerr are nullable. The other fields are required. 

In the object of arrays case, flux must be provided, but some of the elements can be null, and similar for mag and mag err. Is there a straightforward way to capture this in the swagger docs ?

stefanv commented:
I'm not quite sure why, but the spec seems to prohibit arrays with multiples types of items:


> `items` - Value MUST be an object and not an array. Inline or referenced schema MUST be of a Schema Object and not a standard JSON Schema. items MUST be present if the type is array.

stefanv commented:
Ah, good, they fixed this in v3.1 of the spec, since they moved to be compatible with the JSON Schema spec: https://json-schema.org/draft/2019-09/json-schema-core.html#rfc.section.9.3.1.1

stefanv commented:
This is a good site, btw: https://apisyouwonthate.com/

dannygoldstein commented:
OK guys, I implemented the API as requested, please check out the docs here:

https://www.astro.caltech.edu/~danny/static/openapi3.html#/paths/~1api~1photometry/post

All my tests locally are passing, only failures are due to selenium, e.g., 

```E       selenium.common.exceptions.MoveTargetOutOfBoundsException: Message: (1050, 865) is out of bounds of viewport width (1440) and height (803)

/anaconda2/envs/skyportal/lib/python3.6/site-packages/selenium/webdriver/remote/errorhandler.py:242: MoveTargetOutOfBoundsException```

so the review can now be continued

dannygoldstein commented:
I take it back, @acrellin I am getting a single failing test related to the scanning page plot that appears to be more than the usual elementclickintercepted stuff

It appears to be related to bokeh, specifically when I go to the page `test_candidate_date_filtering` is trying to render, I see 

<img width="245" alt="Screen Shot 2020-05-29 at 12 10 27 PM" src="https://user-images.githubusercontent.com/2769632/83281211-c1200b80-a1a5-11ea-957b-eb180bd10320.png">

acrellin commented:
@dannygoldstein maybe a merge issue? Or could any of your changes to plot.py be impacting the scanning page plots?

acrellin commented:
@dannygoldstein I just checked out your branch to do some poking around and am seeing that `make load_demo_data` still fails -- can you rebase or merge latest master again?

dannygoldstein commented:
@acrellin I am unable to reproduce the errors on travis locally, I think they are all related to selenium (and are not real failiures). I went through and ran them by hand and got the expected behavior. and I think for all intents and purposes this is green.

dannygoldstein commented:
Locally: ```
skyportal/tests/frontend/test_groups.py::test_add_new_group_explicit_self_admin PASSED [ 87%]
skyportal/tests/frontend/test_frontpage.py::test_jump_to_page_invalid_values PASSED [ 25%]
```

dannygoldstein commented:
PR is now green.

acrellin commented on skyportal/handlers/api/photometry.py:
```diff
@@ -53,86 +60,130 @@ def post(self):
+            return self.error('Unable to coerce passed JSON to a series of packets. '
+                              f'Error was: "{e}"')
+
+        # pop out thumbnails and process what's left using schemas
```
Are thumbnails being handled here anymore, is that functionality removed?

dannygoldstein replied:
I removed it for separation of concerns, thought that is better handled in the `ThumbnailHandler` endpoint.

acrellin replied:
That's fine

acrellin commented on skyportal/handlers/api/photometry.py:
```diff
@@ -53,86 +60,130 @@ def post(self):
         _ = Source.get_if_owned_by(phot.obj_id, self.current_user)
 
-        return self.success(data=phot)
+        # get the desired output format
```
The below logic should also be applied to `SourcePhotometryHandler` (https://github.com/skyportal/skyportal/blob/94a7c7ba3638a4584afc6d394265e3bccb579ce9/skyportal/handlers/api/source.py#L384) which handles the fetching of all of a source's photometry.  It should be factored out into a function that both this and the other handler call so as to avoid code duplication.  That can be added to this PR or added in an upcoming PR, whichever you prefer.

acrellin replied:
Also, should any of this also apply to the data used in `plot`?

dannygoldstein replied:
The data in `plot` is OK, it's all already standardized (stored in microJanskies).

dannygoldstein replied:
First comment fixed in 3f2936b

acrellin commented on skyportal/schema.py:
```diff
@@ -128,6 +144,368 @@ def add_schema(schema_class_name, exclude=[], add_to_model=False):
             add_schema(f'{schema_class_name}NoID', exclude=['created_at', 'id', 'modified'])
 
 
+class PhotBaseFlexible(object):
```
Can we add a comment or docstring here & below to explain what these different classes are used for? e.g. which are used for generating the docs and which are used for validating data

dannygoldstein replied:
Fixed in 9fe649d

dannygoldstein commented on skyportal/enum.py:
```diff
@@ -0,0 +1,20 @@
+import sqlalchemy as sa
```
Done.

dannygoldstein commented:
Removing @stefanv to keep to one of @stefanv and @acrellin  per review, so ready for your review @profjsb

profjsb commented on skyportal/tests/api/test_photometry.py:
```diff
@@ -4,17 +4,21 @@
 
-def test_token_user_post_get_photometry_data(upload_data_token, public_source):
+
+def test_token_user_post_get_photometry_data(upload_data_token, public_source,
```
Should be including tests for `ra` and `dec` (fail when not provided, pass when given as null, etc.).

dannygoldstein commented on skyportal/models.py:
```diff
@@ -365,33 +359,24 @@ class Comment(Base):
+                                  'the photometric aperture [deg].')
+
+    packet = sa.Column(JSONB)
     altdata = sa.Column(JSONB)
```
Im actually not totally sure what this is supposed to be

acrellin replied:
Pretty much everything else - gaia, simbad, tns, etc info - if necessary, let's rename this in another PR

dannygoldstein replied:
Deferring to another PR

profjsb commented on skyportal/tests/data/phot.csv:
```diff
@@ -1,4 +1,4 @@
-lim_mag,filter,mjd,flux,zp,zpsys,fluxerr
+lim_mag,filter,mjd,flux,zp,magsys,fluxerr
```
include position info and uncertainty

dannygoldstein commented on skyportal/plot.py:
```diff
@@ -17,15 +17,14 @@
-DEFAULT_ZP = 8.9 + 15  # so that things are in muJy
 
 SPEC_LINES = {
     'H': ([3970, 4102, 4341, 4861, 6563], '#ff0000'),
```
Outside the scope of this PR, but created a story on clubhouse

dannygoldstein commented on skyportal/models.py:
```diff
@@ -365,33 +359,24 @@ class Comment(Base):
+    dec = sa.Column(sa.Float, doc='ICRS Declination of the centroid of '
+                                  'the photometric aperture [deg].')
+
+    packet = sa.Column(JSONB)
```
`passed_as`?

dannygoldstein replied:
`original_data`?

dannygoldstein replied:
`user_posted`?

dannygoldstein replied:
renamed to `original_user_data` and added docstring in e4b8c10

dannygoldstein commented on skyportal/models.py:
```diff
@@ -479,9 +464,7 @@ def from_ascii(cls, filename, obj_id, instrument_id, observed_at):
-    type = sa.Column(sa.Enum('new', 'ref', 'sub', 'sdss', 'dr8', "new_gz",
-                             'ref_gz', 'sub_gz',
-                             name='thumbnail_types', validate_strings=True))
+    type = sa.Column(thumbnail_types)
```
Haven't removed it, just bound it to `thumbnail_types` in `phot_enum.py` so it could be used throughout the code

dannygoldstein replied:
Added docstring in 73f26ee

dannygoldstein commented:
@profjsb I think I've addressed your review. If not, please let me know and I'll address any additional issues. Otherwise, the build is passing and I think this is ready to merge

dannygoldstein commented:
> A few more thoughts here inline. I wonder if we should implement a rejection of the photometry when position is given that is very different than that in `Obj`. Could be a good way to protect against errant uploads of the wrong source.

What about for non-sidereal objects?

acrellin commented on skyportal/tests/api/test_photometry.py:
```diff
@@ -92,35 +158,162 @@ def test_token_user_update_photometry(upload_data_token,
+                       data={'obj_id': str(public_source.id),
+                             'mjd': 58000.,
+                             'instrument_id': ztf_camera.id,
+                             'mag': 21.,
```
What happens here in this case with a scalar mag value and list of errors?

dannygoldstein replied:
The mag scalar is broadcast  to match the list of errors. Think pandas
broadcasting logic.

dannygoldstein replied:
The API docs describe the broadcasting rules in detail

acrellin commented:
@danny is there a test for the SourcePhotomtryHandler? Would be good to have, and that being changed here brings it within the scope of this PR

dannygoldstein commented:
The mag scalar is broadcast  to match the list of errors. Think pandas
broadcasting logic.

On Sat, May 30, 2020 at 12:20 PM Ari Crellin-Quick <notifications@github.com>
wrote:

> *@acrellin* commented on this pull request.
> ------------------------------
>
> In skyportal/tests/api/test_photometry.py
> <https://github.com/skyportal/skyportal/pull/472#discussion_r432867441>:
>
> > +    assert status == 200
> +    assert data['status'] == 'success'
> +
> +    np.testing.assert_allclose(data['data']['flux'], 10**(-0.4 * (21 - correction - 23.9)))
> +    np.testing.assert_allclose(data['data']['fluxerr'], 0.2 / (2.5 / np.log(10)) * data['data']['flux'])
> +    np.testing.assert_allclose(data['data']['zp'], 23.9)
> +
> +
> +def test_token_user_mixed_photometry_post(upload_data_token, public_source,
> +                                          ztf_camera):
> +
> +    status, data = api('POST', 'photometry',
> +                       data={'obj_id': str(public_source.id),
> +                             'mjd': 58000.,
> +                             'instrument_id': ztf_camera.id,
> +                             'mag': 21.,
>
> What happens here in this case with a scalar mag value and list of errors?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/skyportal/skyportal/pull/472#pullrequestreview-421425388>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAVEFYD4FKMDGDSI3IQAPLDRUEW3HANCNFSM4NMXMPJA>
> .
>


-- 
Danny Goldstein
http://astro.caltech.edu/~danny

dannygoldstein commented:
> @danny is there a test for the SourcePhotometryHandler? Would be good to have, and that being changed here brings it within the scope of this PR

Fixed in 88bbaee

acrellin commented:
Great, looks good

On Sat, May 30, 2020, 10:37 AM Danny Goldstein <notifications@github.com>
wrote:

> *@dannygoldstein* commented on this pull request.
> ------------------------------
>
> In skyportal/tests/api/test_photometry.py
> <https://github.com/skyportal/skyportal/pull/472#discussion_r432872790>:
>
> > +    assert status == 200
> +    assert data['status'] == 'success'
> +
> +    np.testing.assert_allclose(data['data']['flux'], 10**(-0.4 * (21 - correction - 23.9)))
> +    np.testing.assert_allclose(data['data']['fluxerr'], 0.2 / (2.5 / np.log(10)) * data['data']['flux'])
> +    np.testing.assert_allclose(data['data']['zp'], 23.9)
> +
> +
> +def test_token_user_mixed_photometry_post(upload_data_token, public_source,
> +                                          ztf_camera):
> +
> +    status, data = api('POST', 'photometry',
> +                       data={'obj_id': str(public_source.id),
> +                             'mjd': 58000.,
> +                             'instrument_id': ztf_camera.id,
> +                             'mag': 21.,
>
> The API docs describe the broadcasting rules in detail
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/skyportal/skyportal/pull/472#discussion_r432872790>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABXFGTOVFSKYIHRWMJJOF6DRUE76VANCNFSM4NMXMPJA>
> .
>

dannygoldstein commented:
For what its worth @stefanv @acrellin , you guys were right, this is a much nicer API.

profjsb commented on skyportal/handlers/api/photometry.py:
```diff
@@ -53,86 +60,130 @@ def post(self):
-              application/json:
-                schema: Error
-        """
+        # Moved to bottom of page as f-string
```
Was this change implemented?

dannygoldstein replied:
yes

acrellin commented:
:rocket: :sparkles: Typically we do a squash-merge for these many-commit PRs, especially with how many of these are little clean-up / minor change commits. I'll just do an interactive rebase to squash some of them to clean up the commit history.
